var/spack/repos/builtin/packages/kokkos/package.py

```diff
@@ -2,231 +2,276 @@
+        'pascal60': [False, 'Optimize for the PASCAL60 architecture'],
+        'pascal61': [False, 'Optimize for the PASCAL61 architecture'],
+        'power7': [False, 'Optimize for the POWER7 architecture'],
+        'power8': [False, 'Optimize for the POWER8 architecture'],
+        'power9': [False, 'Optimize for the POWER9 architecture'],
+        'ryzen': [False, 'Optimize for the RYZEN architecture'],
+        'skx': [False, 'Optimize for the Skylake architecture'],
+        'snb': [False, 'Optimize for the Sandybridge architecture'],
+        'turing75': [False, 'Optimize for the TURING75 architecture'],
+        'vega': [False, 'Optimize for the VEGA architecture'],
+        'volta70': [False, 'Optimize for the VOLTA70 architecture'],
+        'volta72': [False, 'Optimize for the VOLTA72 architecture'],
+        'wsm': [False, 'Optimize for the Westmere architecture'],
+    }
+
+    spack_micro_arch_map = {
+        "aarch64": "",
+        "arm": "",
+        "ppc": "",
+        "ppc64": "",
+        "ppc64le": "",
+        "ppcle": "",
+        "sparc": None,
+        "sparc64": None,
+        "x86": "",
+        "x86_64": "",
+        "thunderx2": "THUNDERX2",
+        "k10": None,
+        "zen": "RYZEN",
+        "bulldozer": "",
+        "piledriver": "",
+        "zen2": "RYZEN",
+        "steamroller": "",
+        "excavator": "",
+        "a64fx": "",
+        "power7": "POWER7",
+        "power8": "POWER8",
+        "power9": "POWER9",
+        "power8le": "POWER8",
+        "power9le": "POWER9",
+        "i686": None,
+        "pentium2": None,
+        "pentium3": None,
+        "pentium4": None,
+        "prescott": None,
+        "nocona": None,
+        "nehalem": None,
+        "sandybridge": "SNB",
+        "haswell": "HSW",
+        "mic_knl": "KNL",
+        "cannonlake": "SKX",
+        "cascadelake": "SKX",
+        "westmere": "WSM",
+        "core2": None,
+        "ivybridge": "SNB",
+        "broadwell": "BDW",
+        "skylake": "SKX",
+        "icelake": "SKX",
+        "skylake_avx512": "SKX",
+    }
+
+    spack_cuda_arch_map = {
+        "30": 'kepler30',
+        "32": 'kepler32',
+        "35": 'kepler35',
+        "37": 'kepler37',
+        "50": 'maxwell50',
+        "52": 'maxwell52',
+        "53": 'maxwell53',
+        "60": 'pascal60',
+        "61": 'pascal61',
+        "70": 'volta70',
+        "72": 'volta72',
+        "75": 'turing75',
+    }
+    cuda_arches = spack_cuda_arch_map.values()
+
+    arch_values = list(arch_variants.keys())
+    allowed_arch_values = arch_values[:]
+    for arch in arch_values:
+        for cuda_arch in cuda_arches:
+            if cuda_arch in arch:
+                conflicts("+%s" % arch, when="~cuda",
+                          msg="Must specify +cuda for CUDA backend to use "
+                              "GPU architecture %s" % arch)
+        dflt, desc = arch_variants[arch]
+        variant(arch, default=dflt, description=desc)
```

Comment:
It seems to me that having "host" arch variants is redundant with the arch provided by Spack, and I think Spack takes on the responsibility of being able to cross compile to a different backend arch from the front end. In fact, it looks like down on line 252 this is done, but then the variants are also applied.

Maybe having all these variants is the right way to do it, but I would like to see some discussion on whether the host variants can be eliminated in favor of just mapping spack arch's to the correct kokkos settings.

Reply:
There are a couple reasons. 
1. We may do more with the arch information than just add flags, in which case I need to give Kokkos the expect arch name
2. Not trusting Spack to always do the right thing

Reply:
For 1) that's fine - I propose mapping spack arch to `KOKKOS_ARCH_*`.

For 2), I guess I'd be curious if there are specific bugs... Maybe they could be fixed.

Reply:
I think having variants and spack arch will just be confusing for the user. I would map the spack arch as well.

Reply:
1) is already done, I think? Or did you mean something else? 
I've been very conflicted on all of this. I don't want to add a kokkos-specific variant that already exists "natively" in Spack.  At the same time, I don't expect Kokkos and Spack to always have the exact same list of arches.

Reply:
That's what I was trying to say, perhaps poorly, in my comment. (Roughly) I want you to keep that bit, but delete this line for host-archs:
```
        self.append_args("ARCH",   self.arch_values, options)
```

I know that spack cares a lot of about getting the micro-arch mapping to compiler flags 100% right, and I've seen a lot of evidence of this - maybe @adamjstewart can comment on the work on this front.

Reply:
ok, I see it now, it is in `spack_micro_arch_map`.

Reply:
So I went back and looked at our architecture list. The following are not covered by Spack as either a CUDA or Spack microarchitecture.
````
        'epyc': [False, 'Optimize for the EPYC architecture'],
        'fiji': [False, 'Optimize for the FIJI architecture'],
        'gfx901': [False, 'Optimize for the GFX901 architecture'],
        'vega900': [False, 'Optimize for the VEGA900 architecture'],
        'vega906': [False, 'Optimize for the VEGA906 architecture'],
````
I'm not sure why we use Ryzen and Epyc as opposed to Zen and Zen2 - so potentially we could move Epyc off this list. 

The rest are AMD GPU architectures, which I don't believe is covered by Spack at all, unless someone knows better.  I guess I could make a`gpu_arch` variant that takes those values. 

That would leave microarchitecture (target=), cuda_arch, and gpu_arch as the way to set architecture options. Does anyone have an opinion on this?

Reply:
Okay, I went ahead and just an amd_gpu_arch option. It's sort of pointless, anyway, since there's no hipcc support in Spack yet.